plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go

```diff
@@ -0,0 +1,504 @@
+/**
+ * Copyright (c) 2018 Cisco Systems
+ * Author: Steven Barth <stbarth@cisco.com>
+ */
+
+package cisco_telemetry_mdt
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	internaltls "github.com/influxdata/telegraf/internal/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt/ems"
+	dialout "github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt/mdt_dialout"
+	"github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt/telemetry"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	// Maximum telemetry payload size (in bytes) to accept for GRPC dialout transport
+	tcpMaxMsgLen uint32 = 1024 * 1024
+
+	// IOS XR EMS dialin telemetry GPBKV encoding
+	grpcEncodeGPBKV int64 = 3
+)
+
+// CiscoTelemetryMDT plugin for IOS XR, IOS XE and NXOS platforms
+type CiscoTelemetryMDT struct {
+	// Common configuration
+	Transport      string
+	ServiceAddress string `toml:"service_address"`
+
+	// GRPC dialin settings
+	Username     string
+	Password     string
+	Subscription string
+	Redial       internal.Duration
+	MaxMsgSize   int `toml:"max_msg_size"`
+
+	// GRPC TLS settings
+	TLS bool
+	internaltls.ServerConfig
```

Comment:
It is diffierentiated by the transport type. dialin transport uses ClientConfig and dialout transports use ServerConfig (depending on if telegraf acts as a client or server for telemetry).

Reply:
Right, but the toml library that parses it out of the config will fail as it won't know which struct to assign the values to.

You could add all the settings into `CiscoTelemetryMDT` then where you need it:
```go
type CiscoTelemetryMDT struct {
	TLSCA              string   `toml:"tls_ca"`
	TLSCert            string   `toml:"tls_cert"`
	TLSKey             string   `toml:"tls_key"`
	InsecureSkipVerify bool     `toml:"insecure_skip_verify"`
	TLSAllowedCACerts  []string `toml:"tls_allowed_cacerts"`
...
```

```go
tlsConfig, err := internaltls.ServerConfig{
	TLSCert: c.TLSCert
	TLSKey: c.TLSKey
	TLSAllowedCACerts: c.TLSAllowedCACerts
}.TLSConfig()
```

Reply:
Understood, I remember doing some tests in dialout aka. server mode with TLS at some point, but I am not sure if my collegaues tested TLS with client mode so it may have slipped through. I changed the code according to your suggestions. See new commits below.